### PR TITLE
Multi-GPU Support

### PR DESCRIPTION
**Pull request by jw1912**

### CODE REVIEW

**Comment by Disservin** [comment]
this ready to be tested?

**Comment by jw1912** [comment]
@Disservin it works correctly (verified by Viren on 1,2,3,4x 4090s but atm it requires a large batch size to be a good speedup because the gradient accumulation code is basic and suboptimal.

**Comment by Disservin** [comment]
so the interconnect is limiting too much for it to be worth much ?
on my server the pytorch multigpu is no speedup but for vondele it is since it has nvlink between the gpus

**Comment by jw1912** [comment]
@Disservin yes, but if you increase the batch size by several times the transfers become infrequent enough that it is a speedup on normal multigpu.
It would be cool to get a test on a system with nvlink

**Comment by jw1912** [comment]
I will be trying some alternative strategies to reduce the latency of the interconnect.
